zerver/models/realms.py

```diff
@@ -832,17 +832,24 @@ def upload_quota_gb(self) -> Optional[int]:
         if self.custom_upload_quota_gb is not None:
             return self.custom_upload_quota_gb
 
+        if not settings.CORPORATE_ENABLED:
+            return None
+
         plan_type = self.plan_type
-        if plan_type == Realm.PLAN_TYPE_PLUS:
-            return Realm.UPLOAD_QUOTA_STANDARD
-        elif plan_type == Realm.PLAN_TYPE_STANDARD:
-            return Realm.UPLOAD_QUOTA_STANDARD
-        elif plan_type == Realm.PLAN_TYPE_SELF_HOSTED:
+        if plan_type == Realm.PLAN_TYPE_SELF_HOSTED:  # nocoverage
             return None
-        elif plan_type == Realm.PLAN_TYPE_STANDARD_FREE:
-            return Realm.UPLOAD_QUOTA_STANDARD
-        elif plan_type == Realm.PLAN_TYPE_LIMITED:
+        if plan_type == Realm.PLAN_TYPE_LIMITED:
             return Realm.UPLOAD_QUOTA_LIMITED
+        elif plan_type == Realm.PLAN_TYPE_STANDARD_FREE:
+            return Realm.UPLOAD_QUOTA_STANDARD_FREE
+        elif plan_type in [Realm.PLAN_TYPE_STANDARD, Realm.PLAN_TYPE_PLUS]:
+            from corporate.lib.stripe import get_seat_count
+
+            # Paying customers with few users should get a reasonable minimum quota.
+            return min(
+                get_seat_count(self) * settings.UPLOAD_QUOTA_PER_USER_GB,
```

Comment:
Could also easily be done with
```
            current_plan = get_current_plan_by_realm(self)
            assert current_plan is not None
            invoiced_through = current_plan.invoiced_through
            assert invoiced_through is not None
```

and using `invoiced_through.licenses`, but we're not very good at enforcing the `plan.invoice_through is not None` invariant at least in tests, so it feels potentially troublesome to entangle that billing invariant into this kind of `Realm` method? Even though, at least from my reading of the code, in production this invariant probably does hold.

Reply:
I think that would be better; we're going to be doing this query when loading `/` every time, and having that require doing two nontrivial database queries to ask how many users there are feels expensive. I guess this, naively written, might be 3 queries, but they're all obviously cheap ones fetching a single row using an index, rather than things that would potentially scan the whole user table? 

I suppose that issue could be addressed through caching the value? Not sure what's best, all the options are somewhat annoying.

I think any of these options will probably scale OK in practice, given that we only do the queries at all for fully paid customers, which are a small fraction of total `/` requests... but I worry about having a latency problem with `/` that only impacts paid customers.

Reply:
@timabbott I spent some time working through adjusting tests to correctly enforce this invariant when upgrading a realm to standard, so that they wouldn't break; but then I found actualy billing code that temporarily fails to enforce the invariant and calls `do_change_realm_plan_type`, failing and throwing exceptions due to `assert invoiced_through is not None` failing. 

E.g.
```
            if plan.status < CustomerPlan.LIVE_STATUS_THRESHOLD:
                # Tier and usage limit change will happen when plan becomes live.
                self.do_change_plan_type(tier=plan_tier)

                # LicenseLedger entries are way for us to charge customer and track their license usage.
                # So, we should only create these entries for live plans.
                ledger_entry = LicenseLedger.objects.create(
                    plan=plan,
                    is_renewal=True,
                    event_time=event_time,
                    licenses=licenses,
                    licenses_at_next_renewal=licenses,
                )
                plan.invoiced_through = ledger_entry
                plan.save(update_fields=["invoiced_through"])
```

or 

```
                new_plan = CustomerPlan.objects.get(
                    customer=plan.customer,
                    billing_cycle_anchor=plan.end_date,
                    status=CustomerPlan.NEVER_STARTED,
                )
                new_plan.status = CustomerPlan.ACTIVE
                new_plan.save(update_fields=["status"])
                self.do_change_plan_type(tier=new_plan.tier, background_update=True)
                return None, LicenseLedger.objects.create(
                    plan=new_plan,
                    is_renewal=True,
                    event_time=next_billing_cycle,
                    licenses=licenses_at_next_renewal,
                    licenses_at_next_renewal=licenses_at_next_renewal,
                )
```

Reply:
So it feels like it's too easy to break pieces of the billing flows by pursuing this direction... Maybe we should just do caching like you suggested?

Reply:
Yeah, caching seems fine for this.